diff --git a/dist/node/chunks/dep-DZ2tZksn.js b/dist/node/chunks/dep-DZ2tZksn.js
index 763b46fa0a0f32f78be6b253c5300f1c5f1b0c95..963cb398bcbb8485e8224bf2bd69dccaafccd63b 100644
--- a/dist/node/chunks/dep-DZ2tZksn.js
+++ b/dist/node/chunks/dep-DZ2tZksn.js
@@ -7135,7 +7135,6 @@ function terserPlugin(config$2) {
 		},
 		async renderChunk(code, _chunk, outputOptions) {
 			if (config$2.build.minify !== "terser" && !outputOptions.__vite_force_terser__) return null;
-			if (config$2.build.lib && outputOptions.format === "es") return null;
 			worker ||= makeWorker();
 			const terserPath$1 = loadTerserPath(config$2.root);
 			const res = await worker.run(terserPath$1, code, {
@@ -25619,12 +25618,12 @@ var require_http_proxy = __commonJS({ "../../node_modules/.pnpm/http-proxy@1.18.
 	/*!
 	* Caron dimonio, con occhi di bragia
 	* loro accennando, tutte le raccoglie;
-	* batte col remo qualunque s’adagia 
+	* batte col remo qualunque s’adagia
 	*
 	* Charon the demon, with the eyes of glede,
 	* Beckoning to them, collects them all together,
 	* Beats with his oar whoever lags behind
-	*          
+	*
 	*          Dante - The Divine Comedy (Canto III)
 	*/
 	module.exports = require_http_proxy$1();
